### DISCUSSION TITLE
[server routes/h3] Curiosity question about stack trace when creating HTTP errors

### DISCUSSION THREAD

**Discussion by LoganTann** [nuxt/nuxt#10578] [Questions (Nuxt 2)]
Hello, Maybe a dumb question, but I'm creating a server route to fetch an article from supabase.

The function may return HTTP 400 if the request param is not valid, or 404 if the article does not exists

When I build a 400 error, it prints the stack trace in the description field; but not when I throw a 404 error. Do you guys knows why ?

| 400 err | 404 err |
|---|---|
| ![image](https://user-images.githubusercontent.com/28659185/179493215-0c7fab74-e5b2-4012-a805-e59dc69b3c05.png) | ![image](https://user-images.githubusercontent.com/28659185/179493249-00ba6958-14d2-4984-bbb5-674ed61d9fd3.png) |

Here is my code :

```ts
import { SupabaseClient } from "@supabase/supabase-js";

export async function getArticle(
    event: IncomingMessage,
    client: SupabaseClient,
    columns: string = "id"
): Promise<{ data: any[]; articleId: number }> {
    const articleId = Number(event.context.params.id ?? "invalid");
    if (!Number.isInteger(articleId)) {
        throw createError({
            statusCode: 400,
            message: "parameter [id] must be an integer",
        });
    }
    const results = await client
        .from("novel_chapters")
        .select(columns)
        .eq("id", articleId)
        .limit(1);
    if (!results.data || results.data.length === 0) {
        throw createError({
            statusCode: 404,
            message:
                "The article you requested does not exists in the database",
        });
    }
    return { data: results.data, articleId };
}
```

**Comment by micgoe** [comment]
Same here. Is it on purpose? If so, why is that? What area the reasons for it? I don't want to expose any internal stack trace to my clients consuming from my api.

**Comment by AloisSeckar** [comment]
I believe this is done on purpose. The stack message is appended to any error code other than `404`, but **only in dev mode**.

See impl of: https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/core/runtime/nitro/error.ts (row 18)


I checked the behaviour with simple server side GET method  `/server/api/sample.get.ts` like that:

```js
export default defineEventHandler(() => {
  const code = 400 + Math.floor(Math.random() * 10)
  throw createError({
    statusCode: code,
    message: code.toString()
  })
})
```
You can then start your app, call `http://localhost:3000/api/sample/` in browser and rape F5 button to see it yourself. 

If you go with `npm run dev`, stack messages will be added to any non-404 error responses. If you switch to `npm run start`, you'll never get the stack message regardless the status code. So it is safe for production and it won't expose your backend to strangers.
